### ISSUE TITLE
M6 — Losses and optimisers

### ISSUE THREAD

**Issue by harshitagrawal2O** [harshitagrawal2O/tensorkit#6]
**Milestone 6 of 8** — see [`SPEC.md`](../blob/main/SPEC.md) section 4 for the full definition of done.

**CS subject:** ML optimisation (momentum, bias correction) · Numerical stability

**Nobody is working on this.** Comment to claim it. One person per milestone at a time; if a
claim goes quiet for two weeks it is fair game again.

## Scope

`MSELoss`, `CrossEntropyLoss` (fused log-softmax + NLL), `SGD` with momentum, `Adam`, `AdamW`.

## Definition of done

- Cross-entropy on logits of magnitude 1e4 gives finite loss and gradients
- The fused gradient equals `(softmax(logits) - onehot) / N` to 1e-9
- Adam's first step matches a hand-computed reference **with** bias correction; a companion test asserts the uncorrected version fails

## Where it gets interesting

Adam's `t` starts at **1**, not 0. At t=0 both correction terms are zero and you divide by zero; get the off-by-one wrong and the first step is wildly too large, which looks like a loss curve that is 'just a bit unstable'.

## Ground rules

- Do **not** weaken a test to make it pass. If you believe a test is wrong, open a
  [spec-defect issue](../issues/new?template=spec-defect.yml) instead — several tests have
  already been corrected that way.
- Keep the existing docstrings; replace only the `raise NotImplementedError(...)` line.
- `ruff check`, `ruff format --check` and `mypy` must be clean.

Chapter-length background is in [the book](../blob/main/docs/tensorkit-book.pdf).

```bash
python -m pytest -m m6 -q
```
